src/spec-node/featuresCLI/packageCommandImpl.ts

```diff
@@ -0,0 +1,84 @@
+import path from 'path';
+import tar from 'tar';
+import { Feature } from '../../spec-configuration/containerFeaturesConfiguration';
+import { LogLevel } from '../../spec-utils/log';
+import { isLocalFile, readLocalDir, readLocalFile, writeLocalFile } from '../../spec-utils/pfs';
+import { FeaturesPackageCommandInput } from './package';
+
+
+export interface SourceInformation {
+    source: string;
+    owner?: string;
+    repo?: string;
+    tag?: string;
+    ref?: string;
+    sha?: string;
+}
+
+export interface DevContainerCollectionMetadata {
+    sourceInformation: SourceInformation;
+    features: Feature[];
+}
+
+
+export async function doFeaturesPackageCommand(args: FeaturesPackageCommandInput): Promise<number> {
+    const { output } = args;
+
+    // For each feature, package each feature and write to 'outputDir/{f}.tgz'
+    // Returns an array of feature metadata from each processed feature
+    const metadataOutput = await getFeaturesAndPackage(args);
+
+    if (!metadataOutput) {
+        output.write('Failed to package features', LogLevel.Error);
+        return 1;
+    }
+
+    const collection: DevContainerCollectionMetadata = {
+        sourceInformation: {
+            source: 'devcontainer-cli',
+        },
+        features: metadataOutput,
+    };
+
+    // Write the metadata to a file
+    const metadataOutputPath = path.join(args.outputDir, 'devcontainer-collection.json');
+    await writeLocalFile(metadataOutputPath, JSON.stringify(collection, null, 4));
+
+    return 0;
+}
+
+async function tarDirectory(featureFolder: string, archiveName: string, outputDir: string) {
+    return new Promise<void>((resolve) => resolve(tar.create({ file: path.join(outputDir, archiveName), cwd: featureFolder }, ['.'])));
+}
+
+export async function getFeaturesAndPackage(args: FeaturesPackageCommandInput): Promise<Feature[] | undefined> {
+    const { output, featuresFolder, outputDir } = args;
+    const featureDirs = await readLocalDir(featuresFolder);
```

Comment:
Should we also support a mode where you directly package a single feature directory without pointing to the parent directory and collecting all subdirectories as individual features? I imagine that a lot of people will start with publishing a single feature, so optimizing to make that use case simple would be great.

Reply:
I am not so sure that it is simpler to have two different flags. The interface would then look like (for a single feature in some folder named `src`)

`devcontainer feature package -f ./src/myFeature` 

or

`devcontainer feature package -c ./src`

...both of which would output exactly the same given only one feature in `src`.

If I then add another feature to `./src`, it no longer makes sense to use the former command, since in the spec say features colocated in a repo will be a part of a "collection" and share the same "metadata 


We have the spec structured in a way that we always will generate a metadata artifact, even if there is only a single feature in the 'collection'.  We **wouldn't** want people to run:

`devcontainer feature package -f ./src/myFeature && devcontainer feature package -f ./src/myFeature2`

,since that would generate two `devcontainer-collection.json` files with 1 feature in each.

Reply:
I wasn't thinking that we would need different command line flags. Rather, if users run the command with a directory that points directly to a feature (e.g. it has a `devcontainer-feature.json` and `install.sh` in the directory), then we would package up that feature into a single-feature collection.

Basically I want to keep it simple for repos that are just trying to publish a single feature, which will likely be the majority of user-contributed features. I expect many people will try to run the command from the feature directory, and the command today is expecting a very specific folder structure which features might not all be in.

Reply:
hmm gotcha, that makes sense. I will add that in that auto-detection bit (and perhaps remove the flag, instead opting for a positional argument to supply the path if you aren't invoking from the current directory)

Reply:
addressed this 🚀